docs/architecture-decisions/adr011-plugin-package-structure.md

```diff
@@ -0,0 +1,74 @@
+---
+id: adrs-adr011
+title: ADR011: Plugin Package Structure
+description: Architecture Decision Record (ADR) for Plugin Package Structure
+---
+
+## Context
+
+A core feature of Backstage is the extensibility via plugins. The Backstage
+repository is open for contributions of plugins. Even most of the core features
+are implemented as plugins. A plugin consists of one or multiple packages in the
+`plugins/` directory. Up till now, we have a simple conventions for naming
+plugin packages: Plugins are named `x`, with the option of having a related
+backend plugin called `x-backend` (where `x` is the plugin name, like `catalog`
+or `techdocs`). There is a need for sharing code between the frontend and
+backend of a plugin, between backend plugins, or components and hooks between
+different frontend plugins
+([some examples](https://github.com/backstage/backstage/issues/3655#issuecomment-758166746)).
+This results in emerging plugin packages with shared code, like
+`packages/catalog-client` or `packages/techdocs-common`.
+
+> There is a common prase in software development:
+> [Naming things is hard](https://martinfowler.com/bliki/TwoHardThings.html)
+
+To keep the contributed plugins consistent, this Architecture Decision Record
+provides rules for naming plugin packages.
+
+## Decision
+
+We will place all plugin related code in the `plugins/` directory. The
+`packages/` directory is reserved for core package of Backstage.
+
+We follow this structure for plugin packages (where `x` is the plugin name, for
+example `catalog` or `techdocs`):
+
+- `x`: Contains the main frontend code of the plugin.
+- `x-backend`: Contains the main backend code of the plugin.
+- `x-common-react`: Contains shared widgets, hooks and similar that both the
+  plugin itself (`x`) and third-party frontend plugins can depend on.
+- `x-common-node`: Contains utilities for backends that both the plugin backend
```

Comment:
Thoughts about just calling these `x-react` and `x-node`?

Reply:
They are getting quite long indeed, for me it would be fine to omit the `common-` here.

Reply:
Any other opinions before I change that?

Reply:
Totally ok. They aren't necessarily related as such (in a hierarchical sense) to common anyway. Sounds good

Reply:
Updated, looks like I have to rename some stuff in #4265 😆

Reply:
> Also: `CatalogApi` in `x-common`, and `catalogApiRef` in `x-react` and the latter reexports the former... Or nah? There's a certain level of inconvenience to have to depend on several packages for "no reason" and have several separate import lines etc.
> 
> Is reexporting a bad thing? Opinions?

I think reexport of the API type makes sense in this case, so that you don't have to add a dependency on `-common`. Tbh I think it might be tricky to establish a pattern right now, and look at it case by case. One thing to look out for is whether you can completely remove the need for a package in the app or in plugins by re-exporting some things.